src/ddmd/statementsem.d

```diff
@@ -2490,6 +2489,19 @@ else
                         cs.error("case variables not allowed in final switch statements");
                         errors = true;
                     }
+
+                    /* Also check if the VarExp is declared in a scope outside of this one
+                     * 'sc2' is set to the scope of the switch statement.
+                     */
+                    Scope* sc2 = sc;
+                    while (sc2.enclosing && sc2.enclosing.sw == sw)
+                        sc2 = sc2.enclosing;
+                    if (!sc2.search(cs.exp.loc, v.ident, null))
+                    {
+                        cs.error("case variables cannot be declared in switch bodies");
```

Comment:
Please print the variable being declared, something like:
```
"case variable `%s` declared at %s cannot be declared in switch body", v.toChars(), v.loc.toChars());
```

Reply:
I can't say I'm a fan of putting line numbers in error messages, but ok.

Just as a note, having a separate diagnostic per line is more natural for GDC (LDC too?) as the entire source line is printed and a `^` is inserted at the character position. When at the GNU Cauldron this month, they are wanting to add location ranges also to underline keywords, constants, and expressions (`^~~~`).

Anyway, that's just food for thought right now.